### ISSUE TITLE
ssh: handle SELinux somehow?

### ISSUE THREAD

**Issue by bradfitz** [tailscale/tailscale#4908]
https://twitter.com/dstaley/status/1539773980692733952 has tips about SELinux.

As somebody who knows ~nothing about SELinux, I'm not sure what it's saying. (That's some sort of exception policy you put somewhere?)

/cc @maisem @danderson @raggi

**Comment by danderson** [comment]
If I'm reading the policy violation right, executing a shell seems to be a restricted action for the default "be a daemon" policy on this system - possibly as a defense against RCE trying to spawn a shell?

I don't know enough about SELinux to construct, distribute, install and apply a better policy though.

**Comment by dstaley** [comment]
Yeah, that matches my (also limited!) understanding. I was able to work around it with the following policy:

```
module tailscale 1.0;

require {
        type unconfined_service_t;
        type unconfined_t;
        class process transition;
}

#============= unconfined_service_t ==============
allow unconfined_service_t unconfined_t:process transition;
```

This is likely disabling the specific feature of SELinux that's mitigating against RCE vulnerabilities, so definitely only use this if you know what you're doing!

I have come across [some documentation](https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/8/html/using_selinux/writing-a-custom-selinux-policy_using-selinux) on how to generate a SELinux policy for a given executable, so it's at least theoretically possible to distribute a `tailscale.pp` file that users can compile and install on their own.

**Comment by bradfitz** [comment]
#4914 is a report that this breaks Fedora 36.

It works if you run `setenforce 0` (which obviously isn't an ideal fix).

**Comment by Will-W** [comment]
The cheat here is presumably going to be to copy the SELinux policy chunks that sshd has, given that you have exactly the same security constraints.

**Comment by Will-W** [comment]
I had a little dig into the Fedora OpenSSH package, out of curiosity, and it is of course complicated. It's not just a nasty set of policies, but the code itself has awareness of SELinux - presumably because it has to do some fairly dynamic things based on who you are logging into SSH as etc.

What I might look at if I have time this weekend is a slightly better workaround policy than the one above. Rather than saying "all unconfined daemons can execute arbitrary stuff" we can label tailscaled appropriately and then say "tailscaled can execute arbitrary stuff". That at least narrows the hole a bit, and can be a workaround while a "proper" implementation happens.

**Comment by SebDanielsson** [comment]
A better alternative to `setenforce 0`, which doesn't persist across reboots, is to set SELinux to permissive mode in `/etc/selinux/config`.

Here's a one-liner for all of you bumping into the same problem on Fedora Server:

```sh
sed -i 's/SELINUX=enforcing.*/SELINUX=permissive/' /etc/selinux/config
```

**Comment by gnufied** [comment]
Another alernative is to run:

```
# run this immediately after getting ssh access denied error
 sealert -l "*"
```

Do what the instructions says to allow process that access.

**Comment by gebrk** [comment]
I'm no expert on this but I've done some digging and I've got a different interpretation of the issue here. 

As there is no specific configuration for tailscaled in SELinux it runs as an unconfined service which should mean that SELinux does not place any limits on tailscaled. Ordinary user processes also run unconfined but as a different type (unconfined_t vs unconfined_service_t).

```
❯ ps axZ | grep tailscale
system_u:system_r:unconfined_service_t:s0 28926 ? Ssl   0:48 /usr/sbin/tailscaled --state=/var/lib/tailscale/tailscaled.state --socket=/run/tailscale/tailscaled.sock --port 41641
unconfined_u:unconfined_r:unconfined_t:s0-s0:c0.c1023 36688 pts/4 S+   0:00 grep --color=auto tailscale
```
This means there is nothing preventing tailscaled from executing a shell and if you run `tailscale ssh fedoramachine -t bash -l` it will work though it will still be running in the unconfined service context rather than the unconfined user context.

```
❯ ps axZ | grep bash
unconfined_u:unconfined_r:unconfined_t:s0-s0:c0.c1023 1742 ? Ss   0:00 /usr/bin/ssh-agent /bin/sh -c exec -l /bin/bash -c "/usr/bin/startplasma-x11"
system_u:system_r:unconfined_service_t:s0 37595 pts/8 Ssl   0:00 /usr/sbin/tailscaled be-child ssh --uid=1000 --gid=1000 --groups=1000,10,100,978,977,1001 --local-user=...--remote-user=... --remote-ip=... --has-tty=true --tty-name=pts/8 --cmd=/bin/bash -- -c bash -l
```

I'm not clear what the implications of running the shell in the unconfined service context rather than the unconfined user context would be (if any).

I suspect that this becomes an issue as `tailscale ssh`  with no command uses `login` to launch the user shell
https://github.com/tailscale/tailscale/blob/66d7d2549f3b1673f78f8e6eb2573b83bd5f5a4a/ssh/tailssh/incubator.go#L117
and that `login` is SELinux aware and tried to transition to the correct context for the shell and this is then blocked by SELinux.

Since the SELinux rule isn't blocking shell execution by the unconfined service, only the context transition by `login` I don't think the workaround of allowing the unconfined service to unconfined transition (https://github.com/tailscale/tailscale/issues/4908#issuecomment-1163838481 and https://github.com/tailscale/tailscale/issues/4908#issuecomment-1220954775 both have this effect) will open any great hole in SELinux protections, unlike setting SELinux to permissive.

I think a "proper" fix would involve creating a specific tailscale SELinux context and allowing this transition alongside whatever else tailscale needs to work when confined by SELinux.

**Comment by DentonGentry** [comment]
Related: https://github.com/tailscale/tailscale/issues/5622
A specific SELinux context for `tailscaled` would also resolve that.

**Comment by gebrk** [comment]
> Can you try re-using the context used by openssh daemon for tailscaled?
> 
> ```
> $ chcon --reference=/usr/sbin/sshd /path/to/tailscaled
> $ systemctl restart ...
> ```

After this selinux is preventing tailscale from connecting. Some of the errors I think are due to the labels on the tailscale related files not matching the labels on sshd related files but other seem to be related to selinux restraints on sshd.

I tried to generate a suitable local policy to test get tailscale to connect to allow testing further but couldn't get it working. There was a massive number of logs while testing which makes me doubt whether running in permissive mode would allow us to see if the original transition would still be blocked amongst the noise.

**Comment by queeup** [comment]
Any progress about this issue?

**Comment by bradfitz** [comment]
We need an SELinux expert at this point. We know how to reproduce ourselves.

I think @DentonGentry was trying to find somebody.

**Comment by uber6** [comment]
So I found a temporary solution that seems to work. I asked around and I got a response that seems to fix this.

```
sudo dnf install rpm-build policycoreutils-devel
mkdir /tmp/tailscale
cd /tmp/tailscale
sepolicy generate --init /usr/sbin/tailscaled
sudo ./tailscaled.sh
```

The explanation I got was that the above command will label tailscaled and then also set the policy on it to be permissive.

**Comment by abhiseksanyal** [comment]
I have some experience authoring SELinux policies. I have published a SELinux policy [here](https://github.com/abhiseksanyal/tailscale-selinux-policy)

It does the following,
1. Adds the context "tailscaled_t" for Tailscale service and runs it confined
2. Support Tailscale SSH on systems where SELinux is running in enforced mode

To support Tailscale SSH, it does the following,
1. Adds support to execute login binary when SSH is done
2. Adds support to transition to users that are part of unconfined domain

Tested on Fedora 37 using Tailscale 1.34.2

**Comment by bradfitz** [comment]
@abhiseksanyal, wow, thanks!

Would you like us to take over maintenance of that in the main Tailscale repo and apply those policies by default when installing the Tailscale package on SELinux systems? (The license you picked looks compatible, but wanted to clarify whether that's your intention)

If so, would you like to help? :)

Would you like to thank you gifts either way? :) Email my username at tailscale.com and we'll hook you up with some Tailscale swag.

**Comment by abhiseksanyal** [comment]
> Would you like us to take over maintenance of that in the main Tailscale repo and apply those policies by default when installing the Tailscale package on SELinux systems? (The license you picked looks compatible, but wanted to clarify whether that's your intention)
> 
> If so, would you like to help? :)

I intentionally chose a compatible license so that the policies could be added to the main repo itself
I can help with this.

**Comment by bronson** [comment]
My workaround is to add `-t bash` to the end of the ssh invocation:

```txt
$ ssh larn
Last login: Sat Apr  1 03:59:09 from 100.72.28.60
Fedora CoreOS 37.20230303.3.0
bronson: no shell: Permission denied
Connection to larn closed.

$ ssh larn -t bash
[larn ~]$ 
```

This is adequate for my needs while a good SELinux fix is developed.

**Comment by eliezerlp** [comment]
> I have some experience authoring SELinux policies. I have published a SELinux policy [here](https://github.com/abhiseksanyal/tailscale-selinux-policy)

I noticed a Health check warning after applying these policies on Tailscale 1.52.1 on a fresh Rocky 9.2 VM:
```
# Health check:
#     - dns-os: writing to "/etc/resolv.pre-tailscale-backup.conf" in rename of "/etc/resolv.conf": open /etc/resolv.pre-tailscale-backup.conf: permission denied
#     - dns: writing to "/etc/resolv.pre-tailscale-backup.conf" in rename of "/etc/resolv.conf": open /etc/resolv.pre-tailscale-backup.conf: permission denied
```

**Comment by mochaaP** [comment]
<https://github.com/abhiseksanyal/tailscale-selinux-policy/issues/4#issuecomment-1853921432>

It might be helpful to transition to an unconfined domain when calling `dropPrivileges` in `ssh/tailssh/incubator.go`.

**Comment by bigon** [comment]
Hello,

It seems that when login with SSH, the context of the use is not OK

```
# id -Z
system_u:system_r:unconfined_service_t:s0
```

This is quite a security issue for people needing SELinux

I see there was a [proposed SELinux policy](https://github.com/tailscale/tailscale/issues/4908#issuecomment-1400254802), but that doesn't seems to be integrated in the RPM's?

Anyplan to work on this?

**Comment by compuguy** [comment]
> Hello,
> 
> It seems that when login with SSH, the context of the use is not OK
> 
> ```
> # id -Z
> system_u:system_r:unconfined_service_t:s0
> ```
> 
> This is quite a security issue for people needing SELinux
> 
> I see there was a [proposed SELinux policy](https://github.com/tailscale/tailscale/issues/4908#issuecomment-1400254802), but that doesn't seems to be integrated in the RPM's?
> 
> Anyplan to work on this?

Not sure. Looks like someone forked it and fixed the SSH issue with the policy: https://github.com/korewaChino/tailscale-selinux-policy

**Comment by bigon** [comment]
Best would be that this SELinux policy is merged in the [refpolicy](https://github.com/SELinuxProject/refpolicy) I guess...

**Comment by compuguy** [comment]
> Best would be that this SELinux policy is merged in the [refpolicy](https://github.com/SELinuxProject/refpolicy) I guess...

True, but even that updated policy breaks DNS resolution. Tailscale can't swap/modify `/etc/resolv.conf`.

**Comment by xrisk** [comment]
FWIW, Tailscale ssh works out of the box for me on Rocky Linux 9. I have Tailscale installed via the fedora repo at https://pkgs.tailscale.com/stable/fedora/aarch64 

Bringing this up because of the new health check that’s showing up since I update to 1.80.2

```
# Health check:
#     - SELinux is enabled; Tailscale SSH may not work. See https://tailscale.com/s/ssh-selinux
```

**Comment by rexackermann** [comment]
> FWIW, Tailscale ssh works out of the box for me on Rocky Linux 9. I have Tailscale installed via the fedora repo at https://pkgs.tailscale.com/stable/fedora/aarch64
> 
> Bringing this up because of the new health check that’s showing up since I update to 1.80.2
> 
> ```
> # Health check:
> #     - SELinux is enabled; Tailscale SSH may not work. See https://tailscale.com/s/ssh-selinux
> ```

Same here on Fedora 42

**Comment by bigon** [comment]
@xrisk @rexackermann Did you check under which context your user is logged-in? (`id -Z` should show you that)

Because that's the main issue here IIRC

**Comment by ejectum** [comment]
> > FWIW, Tailscale ssh works out of the box for me on Rocky Linux 9. I have Tailscale installed via the fedora repo at https://pkgs.tailscale.com/stable/fedora/aarch64
> > Bringing this up because of the new health check that’s showing up since I update to 1.80.2
> > ```
> > # Health check:
> > #     - SELinux is enabled; Tailscale SSH may not work. See https://tailscale.com/s/ssh-selinux
> > ```
> 
> Same here on Fedora 42

Same here, tailscale ssh worked right out of the box :)

Fresh installation of oracle linux 9 on a OCI VM. I just followed the [wiki guidelines](https://tailscale.com/docs/install/cloud/oracle-cloud). 

`[opc@worker ~]$ id -Z`
`system_u:system_r:unconfined_service_t:s0`

`[opc@worker ~]$ uname -a`
`Linux worker 6.12.0-200.74.27.2.el9uek.aarch64 #1 SMP Thu Apr  9 01:16:37 PDT 2026 aarch64 aarch64 aarch64 GNU/Linux`

**Comment by bigon** [comment]
> `[opc@worker ~]$ id -Z`
> `system_u:system_r:unconfined_service_t:s0`

That's actually **NOT** correct

It should be `unconfined_u:unconfined_r:unconfined_t:s0-s0:c0.c1023`

**Comment by ejectum** [comment]
> > `[opc@worker ~]$ id -Z`
> > `system_u:system_r:unconfined_service_t:s0`
> 
> That's actually **NOT** correct
> 
> It should be `unconfined_u:unconfined_r:unconfined_t:s0-s0:c0.c1023`

@bigon, no idea why it returns that, i have no experience with oracle linux. 

This is on a fresh installation of ol9 on a vm spinned at OCI. Tailscale installed with the typical script `curl -fsSL https://tailscale.com/install.sh | sh`


<img width="1755" height="788" alt="Image" src="https://github.com/user-attachments/assets/eac71af2-2eea-4464-a2b2-3cc92522c1ff" />

**Comment by bigon** [comment]
@ejectum this is more than probably not a problem of your installation, just the bug not being fixed.

I'm personally not using tailscale with new $CUSTOMER but just a bit sad that this bug is still not fixed after more than 3 years.

Almost all enterprise distributions are using SELinux by default...

**Comment by ChubbyBoySlim** [comment]
> Best would be that this SELinux policy is merged in the [refpolicy](https://github.com/SELinuxProject/refpolicy) I guess...

This!
